tests: Test extern'd globals on MacOS with the Apple Archiver

xclaesse opened the pull request:
This is a regression test for:
https://github.com/mesonbuild/meson/issues/11165.

The test was initially merged as part of:
https://github.com/mesonbuild/meson/pull/10628, but had to be reverted.

A second attempt at fixing the root cause also had to be reverted: https://github.com/mesonbuild/meson/pull/10699.

A 3rd attempt got merged but missed this unit test: https://github.com/mesonbuild/meson/pull/11742.

jpakkane commented:
```
 error: Could not read ffac597fdb6ce8236b5e2da77ab6fbae3d9d9cb7
fatal: revision walk setup failed
error: https://github.com/Homebrew/homebrew-cask did not send all necessary objects

error: Could not read 623226b77a148708197a10c0d1eab3f8d6ca005b
fatal: revision walk setup failed
error: https://github.com/Homebrew/homebrew-core did not send all necessary objects
```

jpakkane commented:
After restart:

```
@(#)PROGRAM:ld  PROJECT:ld64-820.1
BUILD 20:07:01 Nov  7 2022
configured to support archs: armv6 armv7 armv7s arm64 arm64e arm64_32 i386 x86_64 x86_64h armv6m armv7k armv7m armv7em
Library search paths:
	/usr/local/lib
	/Applications/Xcode_14.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk/usr/lib
Framework search paths:
	/Applications/Xcode_14.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk/System/Library/Frameworks/
Undefined symbols for architecture x86_64:
  "_main", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

xclaesse commented:
I think the actual bug is a bit further:
```
[5/5] cc  -o prog prog.p/prog.c.o -L/usr/local/lib -I/usr/local/include -Wl,-dead_strip_dylibs -Wl,-headerpad_max_install_names -Wl,-undefined,error libmylib.a
FAILED: prog 
cc  -o prog prog.p/prog.c.o -L/usr/local/lib -I/usr/local/include -Wl,-dead_strip_dylibs -Wl,-headerpad_max_install_names -Wl,-undefined,error libmylib.a
Undefined symbols for architecture x86_64:
  "_l2", referenced from:
      _l1 in libmylib.a(libfile.c.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

So we did not fix that issue afterall...

CC @ailin-nemui

osalbahr commented:
The test failures at Homebrew/homebrew-core#135874 (on macOS and Linux) might help with further debugging this error.

xclaesse commented:
Added back the original fix from https://github.com/mesonbuild/meson/pull/10628. If I understand correctly that fix has always been needed but was broken because of other bugs that are now fixed. Let's see what CI thinkgs about it...

eli-schwartz commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -36,7 +36,7 @@
 from ..arglist import CompilerArgs
 from ..compilers import Compiler
-from ..linkers import ArLikeLinker, RSPFileSyntax
+from ..linkers import ArLikeLinker, RSPFileSyntax, AppleArLinker
```
Since the time this commit was originally introduced, we now check `static_linker.id == 'applear'` or similar logic. The hidden impl now causes an import error if you try to import it directly.

xclaesse replied:
fixed

xclaesse commented:
macos CI passed, let's merge then.
